contracts/solidity/eligibility/NFTXMerkleEligibility.sol

```diff
@@ -0,0 +1,239 @@
+
+
+    /**
+     * @notice The name of our Eligibility Module.
+     *
+     * @return string
+     */
+
+    function name() public pure override virtual returns (string memory) {    
+        return 'MerkleEligibility';
+    }
+
+
+    /**
+     * @notice Confirms that our module has been finalised and won't change.
+     *
+     * @return bool
+     */
+
+    function finalized() public view override virtual returns (bool) {    
+        return true;
+    }
+
+
+    /**
+     * @notice The address of our token asset contract.
+     *
+     * @return address 
+     */
+
+   function targetAsset() public pure override virtual returns (address) {
+        return 0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85;
+    }
+
+
+    /**
+     * @notice Allow our eligibility module to be initialised with optional
+     * config data.
+     * 
+     * @param configData Encoded config data
+     */
+
+    function __NFTXEligibility_init_bytes(bytes memory configData) public override virtual initializer {
+        (bytes32 _merkleRoot) = abi.decode(configData, (bytes32));
+        __NFTXEligibility_init(_merkleRoot);
+    }
+
+
+    /**
+     * @notice Parameters here should mirror the config struct.
+     * 
+     * @param _merkleRoot The root of our merkle tree
+     */
+
+    function __NFTXEligibility_init(bytes32 _merkleRoot) public initializer {
+        merkleRoot = _merkleRoot;
+        emit NFTXEligibilityInit(_merkleRoot);
+    }
+
+
+    /**
+     * @notice Checks if a supplied token is eligible, which is defined by our merkle
+     * tree root assigned at initialisation.
+     * 
+     * @dev This check requires the token to have already been passed to `processToken`.
+     *
+     * @return bool If the tokenId is eligible
+     */
+
+    function _checkIfEligible(uint tokenId) internal view override virtual returns (bool) {
+        return validTokenHashes[keccak256(bytes(Strings.toString(tokenId)))];
+    }
+
+    /**
+     * @notice Checks if the token requires a precursory validation before it can have
+     * it's eligibility determined.
+     * 
+     * @dev If this returns `true`, `processToken` should subsequently be run before
+     * checking the eligibility of the token.
+     * 
+     * @param tokenId The ENS domain token ID
+     *
+     * @return bool If the tokenId requires precursory validation
+     */
+
+    function requiresProcessing(uint tokenId) public view returns (bool) {
+        return !_processedTokenHashes[keccak256(bytes(Strings.toString(tokenId)))];
+    }
+
+
+    /**
+     * @notice This will run a precursory check by encoding the token ID, creating the
+     * token hash, and then checking this against our merkle tree.
+     *
+     * @param tokenId The ENS token ID being validated
+     * @param merkleProof Merkle proof to validate against the tokenId
+     *
+     * @return bool If the token is valid
+     */
+
+    function processToken(uint tokenId, bytes32[] calldata merkleProof) public returns (bool) {
```

Comment:
Suggest switching public call to an array and using an internal call for each `tokenId`.